packages/repository/src/__tests__/integration/repositories/relation.factory.integration.ts

```diff
@@ -203,6 +210,247 @@ describe('BelongsTo relation', () => {
+    };
+    expect(toJSON(persistedLink[0])).to.deepEqual(toJSON(expected));
+  });
+
+  it('can find an instance via through model', async () => {
+    const order = await hasManyThroughRepo.create(
+      {
+        description: 'an order hasManyThrough',
+      },
+      {
+        throughData: {id: 99},
+      },
+    );
+    const notMyOrder = await orderRepo.create({
+      description: "someone else's order desc",
+    });
+
+    const orders = await hasManyThroughRepo.find();
+
+    expect(orders).to.not.containEql(notMyOrder);
+    expect(orders).to.deepEqual([order]);
+  });
+
+  it('can find instances via through models', async () => {
+    const order1 = await hasManyThroughRepo.create(
+      {
+        description: 'group 1',
+      },
+      {
+        throughData: {id: 99},
+      },
+    );
+    const order2 = await hasManyThroughRepo.create(
+      {
+        description: 'group 2',
+      },
+      {
+        throughData: {id: 98},
+      },
+    );
+    const orders = await hasManyThroughRepo.find();
+
+    expect(orders).have.length(2);
+    expect(orders).to.deepEqual([order1, order2]);
+    const group1 = await hasManyThroughRepo.find({
+      where: {description: 'group 1'},
+    });
+    expect(group1).to.deepEqual([order1]);
+  });
+
+  it('can delete an instance then delet the through model', async () => {
+    await hasManyThroughRepo.create(
+      {
+        description: 'customer 1',
+      },
+      {
+        throughData: {id: 98},
+      },
+    );
+    const anotherHasManyThroughRepo = hasManyThroughFactory(
+      existingCustomerId + 1,
+    );
+    const order2 = await anotherHasManyThroughRepo.create(
+      {
+        description: 'customer 2',
+      },
+      {
+        throughData: {id: 99},
+      },
+    );
+    let orders = await orderRepo.find();
+    let links = await customerOrderLinkRepo.find();
+
+    expect(orders).have.length(2);
+    expect(links).have.length(2);
+
+    await hasManyThroughRepo.delete();
+    orders = await orderRepo.find();
+    orders = await orderRepo.find();
+    links = await customerOrderLinkRepo.find();
+
+    expect(orders).have.length(1);
+    expect(links).have.length(1);
+    expect(orders).to.deepEqual([order2]);
+    expect(links[0]).has.property('orderId', order2.id);
+    expect(links[0]).has.property('customerId', existingCustomerId + 1);
+  });
+
+  it('through models get deleted once the related target gets deleted', async () => {
+    const order1 = await hasManyThroughRepo.create(
+      {
+        description: 'customer 1',
+      },
+      {
+        throughData: {id: 98},
+      },
+    );
+    const anotherHasManyThroughRepo = hasManyThroughFactory(
+      existingCustomerId + 1,
+    );
+    const order2 = await anotherHasManyThroughRepo.create(
+      {
+        description: 'customer 2',
+      },
+      {
+        throughData: {id: 99},
+      },
+    );
+    // when order1 gets deleted, this through instance should be deleted too.
```

Comment:
Hmm any reason to have a second order here? We can delete `order1`, then check link with `id: 98` is also gone, to verify the through instance is deleted.

Reply:
You're right, they are not necessary 🙈 I just wanted to make sure that the implementation satisfies [this comment](https://github.com/strongloop/loopback-next/pull/4438/files#r402105078). I can remove it if the test is confusing.

Reply:
I see it makes sure the constraint only applies to that deleted source model. Then the test case is valid 👍 , my simplified one doesn't check that.

> Here are two scenario to consider & cover by tests, I'll use "Source", "Target" and "Through" as model names instead of Customer/Seller/Order.

1. Given through links: `{sourceId: 1, targetId: 1}` and `{sourceId: 2, targetId: 2}`.
   When I delete all models belonging to source id 1,
   then the target model id 1 and the first link is deleted.

2. Given trough links `{sourceId: 1, targetId: 1}`, `{sourceId: 2, targetId: 1}` and `{sourceId: 2, targetId: 2}`. 
  When I delete all models belonging to source id 1, 
  then the target model id 1 and the first two links are deleted. Only `{sourceId: 2, targetId: 2}` link is preserved.


--- 

I think scenario 2 covers scenario 1, I couldn't think of a reason why 2 pass while 1 fails 🤔 .

Reply:
Haha I think 1 is a general case and 2 is just a more complicated case.